File: Typescript/Stacks/LinkedListBaseStack.ts
import { DBLinkedList } from "../LinkedList/DBLinkedList";
import { StackADT } from "./StackADT";

export class LinkedListBaseStack<T> implements StackADT<T> {
  private list: DBLinkedList<T> = new DBLinkedList<T>();

  public push(element: T): void {
    this.list.add(element);
  }
  public pop(): T | null {
    if (this.list.isEmpty()) throw new Error("Empty Stack!");
    return this.list.removeLast();
  }

  public top(): T | null {
    if (this.list.isEmpty()) throw new Error("Empty Stack!");
    return this.list.peekLast();
  }

  public isEmpty(): boolean {
    return this.list.sizeList() === 0;
  }

  public size(): number {
    return this.list.sizeList();
  }

  public [Symbol.iterator](): Iterator<T> {
    return this.list[Symbol.iterator]();
  }
}

File: Java/src/Stack/LinkedListBaseStack.java
package Stack;

import DLinkedList.DLinkedList;

import java.util.EmptyStackException;
import java.util.Iterator;

public class LinkedListBaseStack<T> implements StackADT<T> {
    private final DLinkedList<T> list = new DLinkedList<T>();

    public LinkedListBaseStack() {
    }

    public LinkedListBaseStack(T element) {
        push(element);
    }

    @Override
    public void push(T element) {
        list.addLast(element);
    }

    @Override
    public T pop() {
        if (isEmpty()) throw new EmptyStackException();
        return list.removeLast();
    }

    @Override
    public T top() {
        if (isEmpty()) throw new EmptyStackException();
        return list.peekLast();
    }

    @Override
    public int size() {
        return list.size();
    }

    @Override
    public boolean isEmpty() {
        return size() == 0;
    }

    @Override
    public Iterator<T> iterator() {
        return list.iterator();
    }
}
